cmd/object-handlers.go

```diff
@@ -187,22 +268,30 @@ func (api objectAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Req
 		w.WriteHeader(http.StatusPartialContent)
 	}
 
-	// Reads the object at startOffset and writes to mw.
+	// Reads the object at startOffset and writes to objectWriter.
 	if err = getObject(ctx, bucket, object, startOffset, length, httpWriter, objInfo.ETag); err != nil {
+		if objInfo.IsCompressed() {
+			httpWriter.Close()
+			wg.Wait()
+		} else {
+			defer httpWriter.Close()
+		}
 		if !httpWriter.HasWritten() && !statusCodeWritten { // write error response only if no data or headers has been written to client yet
 			writeErrorResponse(w, toAPIErrorCode(err), r.URL)
 		}
-		httpWriter.Close()
 		return
 	}
-
 	if err = httpWriter.Close(); err != nil {
 		if !httpWriter.HasWritten() && !statusCodeWritten { // write error response only if no data or headers has been written to client yet
 			writeErrorResponse(w, toAPIErrorCode(err), r.URL)
 			return
 		}
 	}
 
+	if objInfo.IsCompressed() {
+		wg.Wait() // Wait till the copy go-routines retire.
```

Comment:
Do not understand why this is needed.

Reply:
So here in Get Handler , there are actually two concurrent routines playing around
1) The de-compression go routine ( where the actual decompression takes place )
2) The `getObject` (object layer thread) , which writes to the pipeWriter

There is a possibility for [2] to win the execution race , ie, getObject successfully writes to the pipeWriter and returns. 

Even though the `httpWriter.Close` ( which releases the decompression go-routine) is executed immediately after the getObject call.  The decompression go-routine may take sometime for its closures , So we have to wait for it . 

If we are not waiting for the `pipeWriter.Close()` happening the go-routine. There can be an EOF exception thrown as shown below. 

<b> Without `wg.Wait()` </b>

![image](https://user-images.githubusercontent.com/5410427/43893297-0fe3beae-9bec-11e8-90fe-4a28c9498e03.png)

<b> With `wg.Wait()` </b>

![image](https://user-images.githubusercontent.com/5410427/43893384-5b8ce646-9bec-11e8-924d-64e66d4115ac.png)

`

Reply:
I think we have a more fundamental problem here. We cannot do range-reads (with the current approach) without reading & skipping the entire object (up to end-offset) -> May cause huge performance penalty.

One reason we have a chunk-approach in encryption (add. to security) is to avoid exactly that problem. I'm afraid we have to do some similar approach for compression. IMO the current concept doesn't meet non-functional requirements (if I understand it correct)